File: Civitas - Java/src/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author juanc
 */
public class JugadorEspeculador extends Jugador {
    private float fianza;
    private static int factorEspeculador = 2;
    
    public JugadorEspeculador(Jugador conversion, float fianza){
        super(conversion);
        this.fianza = fianza;
        for(TituloPropiedad titulo : conversion.getPropiedades())
            titulo.actualizaPropietarioPorConversion(this);
    }
    
    public JugadorEspeculador convertirJugador( int fianza){
        System.out.println("________________________JUGADOR YA CONVERTIDO________________________");
        return this;
    }
    
    @Override
    protected int getCasasMax(){
        return factorEspeculador*CasasMax ;
    }
    
    @Override
    protected int getHotelesMax(){
        return factorEspeculador*HotelesMax ;
    }
    
    @Override
    boolean pagaImpuesto(float cantidad){
       return super.pagaImpuesto(cantidad/2) ;
    }
    
    @Override
    protected boolean puedoEdificarCasa(TituloPropiedad propiedad){
        boolean result=false;
        float precio = propiedad.getPrecioEdificar();
        
        if(puedoGastar(precio) && (propiedad.getNumCasas() < this.getCasasMax()) )
            result = true;
        
        return result; 
    }
    
    @Override
    protected boolean puedoEdificarHotel(TituloPropiedad propiedad){
        boolean puedoEdificarHotell=false;
        float precio = propiedad.getPrecioEdificar();
        
        if(puedoGastar(precio) && (propiedad.getNumHoteles()<this.getHotelesMax()) && (propiedad.getNumCasas()>=getCasasPorHotel()) )
            puedoEdificarHotell = true;
        
        return puedoEdificarHotell;
    }
    
    @Override
    protected boolean debeSerEncarcelado(){
        boolean debeSerEncarcelado = super.debeSerEncarcelado();
        boolean puedoPagar = puedoGastar(this.fianza);
        
        if(puedoPagar)
            super.paga(fianza);    
       
        return !puedoPagar && debeSerEncarcelado;
    }
    
    @Override
    public String toString(){
        String devolver = super.toString();
        devolver = devolver.replaceAll("JUGADOR", "JUGADOR ESPECULADOR");
        devolver += "\n     -Fianza " + this.fianza;
        return devolver;
    }
    
}

File: Civitas - Ruby/lib/jugador_especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Jugador_especulador < Jugador
    @@factor_especulador = 2
    attr_reader :fianza
    
    private_class_method :new
    def initialize(fianza)
      @fianza = fianza
    end
    
    #NECESITA EXPLICACION
    def self.nuevo_especulador(jugador, fianza)
      especulador = new fianza
      especulador.copia(jugador) 
      
      especulador.propiedades.each do |titulo|
        titulo.actualiza_propietario_por_conversion(especulador)
      end
      
      return especulador
    end
    
    def convertir_especulador fianza
      puts "___________________JUGADOR YA CONVERTIDO___________________"
      return self
    end
    
    def paga_impuesto cantidad
      return super(cantidad/2)
    end
    
    def puedo_edificar_casa propiedad
      result=false
      precio = propiedad.precio_edificar
      if( puedo_gastar(precio) && ( propiedad.num_casas < (@@casas_max * @@factor_especulador) ) )
          result = true
      end
      return result;
    end
    
    def puedo_edificar_hotel propiedad
      puedo_edificar_hotell = false
      precio = propiedad.precio_edificar
      
      if( puedo_gastar(precio) && (propiedad.num_hoteles < (@@hoteles_max * @@factor_especulador)) && (propiedad.num_casas >= @@casas_por_hotel) )
          puedo_edificar_hotell = true
      end
      
      return puedo_edificar_hotell
    end
   
    def debe_ser_encarcelado
      debe_ser = super
      #             PREGUNTAR HERENCIA DDE METODOS
      puedo_pagar = puedo_gastar(@fianza)

      if(puedo_pagar)
          #PREGUNTAR
          super.paga(@fianza)   
      end

      return !puedo_pagar && debe_ser
      
    end
    
    def to_s
      devolver = super
      devolver["JUGADOR"] = "JUGADOR ESPECULADOR"
      devolver += "\n     -Fianza #{@fianza}\n"
      return devolver
    end
    
  end
end
